libmscore/scale.cpp

```diff
@@ -0,0 +1,478 @@
+                        to.notes[tpc - TPC_MIN] = QString::number(value -
+                              getDifference(getStandardNoteValue(tpc), getStandardNoteValue(prevTpc)));
+                  else if (to.storingMode == ABSOLUTE_FREQUENCY)
+                        to.notes[tpc - TPC_MIN] = QString::number(exp(((double)value / 1200) * log(2)));
+                  }
+            }
+      }
+
+//---------------------------------------------------------
+//   getTuning
+//---------------------------------------------------------
+
+int Scale::getTuning(const Note* note) const
+      {
+      int tpc = note->tpc1();
+      Q_ASSERT(tpc >= TPC_MIN || tpc <= TPC_MAX);
+      return computedTunings[tpc - TPC_MIN];
+      }
+
+//---------------------------------------------------------
+//   load
+//---------------------------------------------------------
+
+void Scale::read(XmlReader& e)
+      {
+      clear();
+
+      name = e.attribute("name");
+      minTpc = e.intAttribute("minTpc");
+      maxTpc = e.intAttribute("maxTpc");
+      updatePitches = (bool)e.intAttribute("updatePitches");
+
+      int tpc = TPC_MIN;
+      while (e.readNextStartElement()) {
+            const QStringRef& tag(e.name());
+            if (tag == "note") {
+                  if (e.hasAttribute("original"))
+                        originalNotes[tpc - TPC_MIN] = e.attribute("original");
+
+                  computedTunings[tpc - TPC_MIN] = e.readDouble();
+                  ++tpc;
+                  }
+            }
+
+      if (tpc != TPC_MAX + 1) {
+            qDebug() << "Failed to read scale data from file!";
+            clear();
+            }
+
+      for (tpc = minTpc; tpc <= maxTpc; ++tpc) {
+            if (originalNotes[tpc - TPC_MIN] == "") {
+                  qDebug() << "Original notes are missing!";
+                  originalNotes[tpc - TPC_MIN] =
+                        QString::number(computedTunings[tpc - TPC_MIN] +
+                              getStandardNoteValue(tpc));
+                  }
+            }
+      }
+
+//---------------------------------------------------------
+//   save
+//---------------------------------------------------------
+
+void Scale::write(Xml& xml) const
+      {
+      bool defaultScale = true;
+      for (int i = 0; i < TPC_NUM_OF; ++i) {
+            if (computedTunings[i] != 0)
+                  defaultScale = false;
+            }
+      if (defaultScale)
+            return;
+
+      QString startTag = QString("Scale name=\"%1\" minTpc=\"%2\" maxTpc=\"%3\" updatePitches=\"%4\"")
+                         .arg(name).arg(minTpc).arg(maxTpc).arg(updatePitches);
+      xml.stag(startTag);
+
+      for (int i = 0; i < TPC_NUM_OF; ++i) {
+            QString noteTag = "note";
+            if (originalNotes[i] != "")
+                  noteTag = QString("note original=\"%1\"").arg(originalNotes[i]);
+            xml.tag(noteTag, (double)computedTunings[i]);
+            }
```

Comment:
Not sure this would work as (I imagine it is) intended: if `originalNotes` as a value and the `noteTag` var is expanded, I suspect `noteTag` is no longer a valid tag for `Xml::tag()`.